pipeline/common/downloads.py

```diff
@@ -313,3 +316,154 @@ def download_chunks(self) -> Generator[bytes, None, None]:
     def decode(self, byte_stream) -> Generator[bytes, None, None]:
         """Pass through the byte stream. This method can be specialized by child classes."""
         return byte_stream
+
+
+@contextmanager
+def _read_lines_multiple_files(
+    files: list[Union[str, Path]], path_in_archive: Optional[str]
+) -> Generator[str, None, None]:
+    """
+    Iterates through each line in multiple files, combining it into a single stream.
+    """
+
+    def iter(stack: ExitStack):
+        for file_path in files:
+            logger.info(f"Reading lines from: {file_path}")
+            lines = stack.enter_context(read_lines(file_path, path_in_archive))
+            yield from lines
+            stack.close()
+
+    try:
+        stack = ExitStack()
+        yield iter(stack)
+    finally:
+        stack.close()
+
+
+@contextmanager
+def _read_lines_single_file(location: Union[Path, str], path_in_archive: Optional[str] = None):
+    """
+    A smart function to efficiently stream lines from a local or remote file.
+    The location can either be a URL or a local file system path.
+    It handles gzip, zst, and plain text files.
+
+    Args:
+        location - URL or file path
+        path_in_archive  - The path to a file in a zip archive
+    """
+    try:
+        location = str(location)
+        stack = ExitStack()
+
+        if location.startswith("http://") or location.startswith("https://"):
+            # This is a remote file.
+
+            response = requests.head(location, allow_redirects=True)
+            content_type = response.headers.get("Content-Type")
+            if content_type == "application/gzip":
```

Comment:
There are a lof of ifs. Should we cover all them with tests? This applies to the whole file.